Optimize queries to be 20 times faster

jonasseglare opened the pull request:
#### SUMMARY

***Background:*** We noticed that Datahike queries take a lot of time. Using the [clj-async-profiler][100] revelealed that Datahike queries with clauses that have multiple unknown variables, e.g. `[?c :concept/id ?from-id]` take a lot of time because they result in calls to the `datahike.db.interface.IDB/-search` protocol methods where the database backend is asked to find all matches for those variables. However, often it is possible to substitute a variable for a set of known possible values. For instance if we know that `?c` can only be either `119`, `120` or `121`, the pattern `[?c :concept/id ?from-id]` can be replaced by `[119 :concept/id ?from-id]`, `[120 :concept/id ?from-id]` and `[121 :concept/id ?from-id]`.

***Proposed improvement:*** Change the `datahike.query` engine so that it attempts to substitute variables in patterns to make the patterns more specific. This is done by introducing the function `expand-constrained-patterns`. Exactly how the substitution should be performed is determined by a *relprod strategy* in located in the `context` map at `(-> context :settings :relprod-strategy)`. I tried out a few different strategies:

* `identity` (`no-strategy`): Don't perform any substitutions
* `select-simple`: Only perform substitutions that result in no or one pattern.
* `select-all`: Perform all possible substitutions
* `expand-once`: This performs all substitutions of `select-simple` *and one more substitution* (if possible) that results in more patterns.

I ran a set of 135 test requests against our API and noticed that the best strategy `expand-once` is about *20 times faster* than not trying to substitute any variables (`no-strategy`).

![total_time_plot](https://github.com/replikativ/datahike/assets/5850088/27c821e5-6a54-4d64-ab71-b9342703c86c)

The strategies are implemented as the functions `clojure.core/identity`, `relprod-select-simple`, `relprod-select-all` and `relprod-expand-once` and can be passed by including the strategy at the path `[:settings :relprod-strategy]` in the query.

[100]: https://github.com/clojure-goes-fast/clj-async-profiler

jonasseglare commented on src/datahike/query.cljc:
```diff
@@ -517,47 +536,56 @@
-                        (persistent!))]
-        (Relation. (zipmap (concat keep-attrs1 keep-attrs2) (range))
-                   new-tuples)))))
+      (hash-join-tuples params1 params2)
```
**Note**: I though `hash-join` had too much repetition so I refactored it. If you think this change is too unrelated to the rest of the code, I am happy to remove it from the PR.

TimoKramer commented:
Hey @jonasseglare, thanks for submitting this PR. Please give us some time to review this.

whilo commented:
@jonasseglare this makes a lot of sense as an optimization and is a very nice contribution! I have thought about similar optimizations before, but did not get to them. Another one would be to to use forward (and potentially reactivated backward) iterators work for `<` and `<=` predicates to not first fetch everything from the index and then filter. Finally @jsmassa and I looked into query planning (clause reordering), but have not been able to implement it yet (we mostly worked on ensuring that an ordering is runnable, i.e. all variables are bound when they are needed).

In general it would be good to make optimal choices somewhat automatically (or pick good defaults), since it can be hard for the average use case/user to think about the optimizations. I am not sure which of your option would be a safe default, probably `select-simple`, but I will check the implementation details.

jonasseglare commented:
Thank you for looking into this PR! Yes, the idea of letting the strategy be a parameter is to have a safe default but provide opportunities for better performance for those who need it. I believe `select-simple` would be a safe default and an improvement over `identity` which is equivalent to what we currently have, that is no substitutions at all. I have not read up on the literature on this but thought that attempting to substitute variables in clauses would be a fairly simple change that can make a big difference, at least this is what we observe for our use case. And then we can possibly implement something even better later.

I am intrigued by your work on the query planner! How the clauses are ordered is probably related to how variables in the clauses are substituted. Without knowing anything about this topic, I would suppose that processing clauses with few unknown variables first would make it easier to solve clauses with more unknowns later. @whilo Do you have a pointer to a particular query planning algorithm that you plan to implement?

whilo commented:
Sorry, for not being super responsive. I am very excited about this PR, but am currently busy with an important PhD milestone that should be done next week and I also need to merge my other PR for distributed datahike first. 

Here are two fairly relevant papers for query planning https://dl.acm.org/doi/abs/10.1145/3183713.3183733, https://link.springer.com/chapter/10.1007/978-3-031-16767-6_5. Souffle in particular is the most scalable Datalog implementation I know, but it usually gets compiled ahead of time for a fixed dataset and query and then run once (for static analysis, there is an interpreter now as well https://souffle-lang.github.io/interpreter.html). 

The first step is to get join size estimates, the second step would be a form of abstract interpretation of the query where you compute sizes instead of actual results and then search for the most efficient plan. Postgres for instance still seems to do exhaustive search up to a certain join size (12 joins I think) and then uses a weak evolutionary algorithm to optimize larger joins. One good way to retain sizes of the base relations is to use hyperloglog as xtdb does.

 @jonasseglare are you on the Clojurians slack? I think it would be easiest to discuss generally a bit there.

whilo commented:
Now that the server is merged, I would focus on getting this merged before improving the writer latency. @jonasseglare can you remove merge conflicts?

jonasseglare commented:
@whilo Great to hear! I am investigating unit test that started to fail after I rebased. I will write here again once I am done.

jonasseglare commented:
I have rebased this branch on top of the latest changes and solved the conflicts and made all tests pass.

You can move forward with this merge request now if you like :-)

jonasseglare commented on src/datahike/query.cljc:
```diff
@@ -1244,6 +1390,7 @@
       true                                          (-post-process qfind)
       qreturnmaps                                   (convert-to-return-maps qreturnmaps)
       stats?                                        (#(-> context-out
+                                                          (update :settings dissoc :relprod-strategy) ;; <-- Don't return non-serializable data.
```
Maybe we don't need to expose `:settings` at all and can add it to the line below so that it becomes `(dissoc :rels :sources :settings)`.

whilo commented:
@jonasseglare Lmk if my feedback above was helpful. I still don't fully understand the expand-once strategy, but overall it makes sense how you partially evaluate and expand patterns from the tuples that are already known. In general it would be great to document/formalize it a bit, because that will facilitate reasoning, but that is not needed to get it merged for now.

jonasseglare commented:
@whilo Thank you for the review! I think your feedback was very useful, I just haven't had time to address the comments yet.

Regarding the effectiveness of different strategies such as expand-once, I believe it depends a lot on what queries are performed and the shape of the data stored in the database. I still haven't had time to make an easy-to-run benchmark.

whilo commented:
Ok, cool! Let me know if I can help in any way. I still want to go through the expand-once process in more detail. My current mental model tells me that the strategies can be suboptimal compared to the current implementation in the case where the current iterator would fetch a bunch of matches at once, while the strategy translates it into individual patterns and then maps them individually, each time incurring the cost of a lookup compared to a faster batched retrieval by the iterator. Another cost is the translation into clauses itself. What is your sense when strategies become suboptimal? In my current understanding partial evaluation should not be slower than grounding on the whole database, do you think this as well?

jonasseglare commented on src/datahike/query.cljc:
```diff
@@ -87,6 +88,15 @@
 
 ;; Utilities
 
+(defn distinct-tuples [tuples]
```
Good point. The reason for this function is that it seemed that tuples are sometimes represented by Java object arrays (supposedly for performance reasons). And object arrays are equal only if their JVM references are equal. Because this function compares them by the values of applying `vec`, two different object arrays will be considered equal if they contain the same elements. I wrote an explanation in the code and improved the unit test.

jonasseglare commented on src/datahike/query.cljc:
```diff
@@ -972,6 +1003,132 @@
+on the data. 
+
+This might be the strategy used by Datomic,
+which can be seen if we request 'Query stats' from Datomic:
```
No, I haven't taken any time for that but it could be a useful comparison to do at some point.

jonasseglare commented on test/datahike/test/api_test.cljc:
```diff
@@ -873,3 +874,31 @@
+                              [2 :likes "candy"]
+                              [3 :likes "pie"]
+                              [4 :likes "pizza"]}]})))))
+    (is (= [:identity :simple :all :once]
```
Yes, the purpose of this test is to check that when i provide a value of the `:relprod-strategy` setting, I want to make sure that value is actually used when the query is evaluated (and not ignored). But I realized that the way the code is written now is more complicated than it needs to be. I don't think there is a way to get rid of the state completely, but I moved the state *inside* the `doseq` form and replaced it by `strategy-was-called (atom false)` that will be set to `true` if the strategy is called.

jonasseglare commented on src/datahike/query.cljc:
```diff
@@ -972,6 +1003,132 @@
+  (relprod-filter relprod #(<= (:tuple-count %) 1)))
+
+(defn relprod-expand-once
+  "This strategy first performs `relprod-select-simple` and 
```
This particular strategy was just something that I tried out and that happened to lead to shorter evaluation times on the particular benchmark I was running :-) It is a sort of middle-ground/tradeoff between only doing substitutions with relations that contain 1 tuple and doing substitutions with *all tuples*.

So, suppose we have relations with the following tuple counts: `[1 1 1 10 11 12]`.

* `relprod-select-simple` means doing substitutions using the first three relations for which the tuple count is 1. This results in 1 more specific pattern that we use to query the database.
* `relprod-expand-once` means doing substitutions using the first four relations with tuple counts `[1 1 1 10]`. This results in up to 10 more specific patterns that we use to query the database (the product 1*1*1*10).
* `relprod-select-all` would mean that we take do substitutions using all relations and end up with up to 1*1*1*10*11*12 = 1320 more specific patterns.

On the data that I was experimenting on, `relprod-expand-once` seemed to be a good middle ground.

jonasseglare commented on test/datahike/test/api_test.cljc:
```diff
@@ -873,3 +874,31 @@
+                              [:db/add -1 :friend 296]]
+                 :keep-history? false
+                 :schema-flexibility :read}
+            conn (utils/setup-db cfg)]
```
Good point, I removed this code.

I also added a larger test, `test/datahike/test/strategy_test.clj`

jonasseglare commented on src/datahike/query.cljc:
```diff
@@ -972,6 +1003,132 @@
+        :let [mentioned-vars (filter attrs vars)]
+        :when (seq mentioned-vars)]
+    {:rel rel
+     :vars mentioned-vars
```
Good point. No, I can actually remove `:key` here. But to clarify my intent to uniquely identify each item by its `:vars`, I added a function `rel-data-key` that looks up the value at `:vars`.

jonasseglare commented on src/datahike/query.cljc:
```diff
@@ -910,27 +929,39 @@
+       added added)
+     pattern)))
+
+(defn resolve-pattern-lookup-refs-or-nil
```
Actually, it seems like I included this function for dealing with some of the test cases listed in this issue:

https://github.com/replikativ/datahike/issues/650

But because we haven't decided what to do in case a variable is bound to illegal values, maybe it is better to not deal with it. So I simply removed the function `resolve-pattern-lookup-refs-or-nil` and called `resolve-pattern-lookup-refs` instead. I also made other simplifications. This makes the PR more focused.

jonasseglare replied:
OK, I realized I had to undo my previous change and put back the function `resolve-pattern-lookup-refs-or-nil`. I added an explanation in the docstring of the function why it is needed.

jonasseglare commented:
I have addressed all the comments now. Most importantly, I added a file [`strategy_test.clj`](https://github.com/jonasseglare/datahike/blob/optimize-query-patterns/test/datahike/test/strategy_test.clj) where I compare the running times of different strategies on a synthetic example inspired by the JobTech Taxonomy.

Before we merge, are there any cosmetic changes to be made? For instance, are the names of the strategies `relprod-select-all`, `relprod-select-simple` and `relprod-expand-once` good? When we specify a strategy in a query, is `:settings {:relprod-strategy ...}` a good way to pass in the strategy to the query engine or do we want to rename some keys?

whilo commented:
> I have addressed all the comments now. Most importantly, I added a file [`strategy_test.clj`](https://github.com/jonasseglare/datahike/blob/optimize-query-patterns/test/datahike/test/strategy_test.clj) where I compare the running times of different strategies on a synthetic example inspired by the JobTech Taxonomy.
> 
> Before we merge, are there any cosmetic changes to be made? For instance, are the names of the strategies `relprod-select-all`, `relprod-select-simple` and `relprod-expand-once` good? When we specify a strategy in a query, is `:settings {:relprod-strategy ...}` a good way to pass in the strategy to the query engine or do we want to rename some keys?

Thank you! You do not need to put the prefix `relprod-` into the name of each strategy, as it is redundant.

jonasseglare commented on test/datahike/test/strategy_test.clj:
```diff
@@ -0,0 +1,289 @@
+(ns datahike.test.strategy-test
```
Here is an extract of the log output from this test:

```

2023-12-21T12:51:04.405Z jonas-ostlund-jobtech INFO [datahike.test.strategy-test:165] - Query related to 1 concept:
2023-12-21T12:51:04.405Z jonas-ostlund-jobtech INFO [datahike.test.strategy-test:96] -       select-all: 0.000608
2023-12-21T12:51:04.406Z jonas-ostlund-jobtech INFO [datahike.test.strategy-test:96] -    select-simple: 0.000730
2023-12-21T12:51:04.406Z jonas-ostlund-jobtech INFO [datahike.test.strategy-test:96] -      expand-once: 0.000916
2023-12-21T12:51:04.406Z jonas-ostlund-jobtech INFO [datahike.test.strategy-test:96] -         identity: 0.034990
2023-12-21T12:51:04.491Z jonas-ostlund-jobtech INFO [datahike.test.strategy-test:180] - Query related to 2 concepts:
2023-12-21T12:51:04.491Z jonas-ostlund-jobtech INFO [datahike.test.strategy-test:96] -      expand-once: 0.000811
2023-12-21T12:51:04.492Z jonas-ostlund-jobtech INFO [datahike.test.strategy-test:96] -       select-all: 0.000930
2023-12-21T12:51:04.492Z jonas-ostlund-jobtech INFO [datahike.test.strategy-test:96] -    select-simple: 0.036576
2023-12-21T12:51:04.492Z jonas-ostlund-jobtech INFO [datahike.test.strategy-test:96] -         identity: 0.044850

....

2023-12-21T12:51:03.904Z jonas-ostlund-jobtech INFO [datahike.test.strategy-test:240] - Find all edges between a few small sets of concept types
2023-12-21T12:51:03.905Z jonas-ostlund-jobtech INFO [datahike.test.strategy-test:96] -       select-all: 0.001149
2023-12-21T12:51:03.906Z jonas-ostlund-jobtech INFO [datahike.test.strategy-test:96] -      expand-once: 0.010314
2023-12-21T12:51:03.906Z jonas-ostlund-jobtech INFO [datahike.test.strategy-test:96] -    select-simple: 0.037641
2023-12-21T12:51:03.906Z jonas-ostlund-jobtech INFO [datahike.test.strategy-test:96] -         identity: 0.038294

....

2023-12-21T12:51:05.085Z jonas-ostlund-jobtech INFO [datahike.test.strategy-test:287] - Find all edges between concept types:
2023-12-21T12:51:05.085Z jonas-ostlund-jobtech INFO [datahike.test.strategy-test:96] -         identity: 0.007881
2023-12-21T12:51:05.086Z jonas-ostlund-jobtech INFO [datahike.test.strategy-test:96] -      expand-once: 0.009720
2023-12-21T12:51:05.086Z jonas-ostlund-jobtech INFO [datahike.test.strategy-test:96] -    select-simple: 0.025386
2023-12-21T12:51:05.086Z jonas-ostlund-jobtech INFO [datahike.test.strategy-test:96] -       select-all: 0.504329

```

whilo replied:
Given this benchmark `expand-once` would be a much better default.

jonasseglare replied:
I sort of agree that `expand-once` may be a good default: In none of these examples does it perform worse than `identity` or `select-simple`.

There is one case where we have `select-all: 0.000775` and `expand-once: 0.009825` being more than 10 times slower. So at the moment there doesn't seem to be one-size-fits-all strategy. But the measurements on our API seem to suggest that `expand-once` mostly is just as good as `select-all` or even slightly better.

jonasseglare commented:
I made three more changes:

* Put back the function `resolve-pattern-lookup-refs-or-nil`, it is actually needed. I explain why in the docstring.
* Changed the default strategy to `select-all`.
* Renamed the strategies by removing the `relprod-` prefix from their names.
